crates/wasm-encoder/src/reencode.rs

```diff
@@ -0,0 +1,1308 @@
+                    let mut functions = crate::FunctionSection::new();
+                    reencoder.parse_function_section(&mut functions, section)?;
+                    module_section_flush_codes(module, &functions, &mut codes);
+                }
+                wasmparser::Payload::TableSection(section) => {
+                    let mut tables = crate::TableSection::new();
+                    reencoder.parse_table_section(&mut tables, section)?;
+                    module_section_flush_codes(module, &tables, &mut codes);
+                }
+                wasmparser::Payload::MemorySection(section) => {
+                    let mut memories = crate::MemorySection::new();
+                    reencoder.parse_memory_section(&mut memories, section)?;
+                    module_section_flush_codes(module, &memories, &mut codes);
+                }
+                wasmparser::Payload::TagSection(section) => {
+                    let mut tags = crate::TagSection::new();
+                    reencoder.parse_tag_section(&mut tags, section)?;
+                    module_section_flush_codes(module, &tags, &mut codes);
+                }
+                wasmparser::Payload::GlobalSection(section) => {
+                    let mut globals = crate::GlobalSection::new();
+                    reencoder.parse_global_section(&mut globals, section)?;
+                    module_section_flush_codes(module, &globals, &mut codes);
+                }
+                wasmparser::Payload::ExportSection(section) => {
+                    let mut exports = crate::ExportSection::new();
+                    reencoder.parse_export_section(&mut exports, section)?;
+                    module_section_flush_codes(module, &exports, &mut codes);
+                }
+                wasmparser::Payload::StartSection { func, .. } => {
+                    module_section_flush_codes(
+                        module,
+                        &crate::StartSection {
+                            function_index: reencoder.function_index(func),
+                        },
+                        &mut codes,
+                    );
+                }
+                wasmparser::Payload::ElementSection(section) => {
+                    let mut elements = crate::ElementSection::new();
+                    reencoder.parse_element_section(&mut elements, section)?;
+                    module_section_flush_codes(module, &elements, &mut codes);
+                }
+                wasmparser::Payload::DataCountSection { .. } => (),
+                wasmparser::Payload::DataSection(section) => {
+                    let mut data = crate::DataSection::new();
+                    reencoder.parse_data_section(&mut data, section)?;
+                    module_section_flush_codes(module, &data, &mut codes);
+                }
+                wasmparser::Payload::CodeSectionStart { .. } => {
+                    if let Some(codes) = codes.replace(crate::CodeSection::new()) {
```

Comment:
Here it should be ok to assert that `codes` is `None` and I think that this'll want to fold in the sub-iteration loop to eagerly consume all `CodeSectionEntry` payloads next to create the `CodeSection`. That way the flushing of the code section can be localized to here without having to happen in all other arms as well

Reply:
Done